feat: [ISS-2950] IAM auth for AWS redis

willwill96 opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description

Adds configuration which allows Oauth2 Proxy to connect to AWS redis services using IAM auth.

## Motivation and Context

This should be useful for organizations who use AWS IAM auth for redis connections.

Resolves #2950 

## How Has This Been Tested?

I primarily tested this in my organization's infrastructure. I may be able to get permission to screen share the before/after of these changes, if that's useful. Otherwise, I can try to get a setup working in my AWS cluster.


## Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [ ] My change requires a change to the documentation or CHANGELOG.
- [ ] I have updated the documentation/CHANGELOG accordingly.
- [ ] I have created a feature (non-master) branch for my PR.
- [ ] I have written tests for my code changes.

willwill96 commented on pkg/sessions/redis/aws-iam/auth.go:
```diff
@@ -0,0 +1,105 @@
+	if err != nil {
+		return nil, err
+	}
+
+	credentials, err := cfg.Credentials.Retrieve(ctx)
+
+	if err != nil {
+		return nil, err
+	}
+
+	if credentials.AccessKeyID == "" || credentials.SecretAccessKey == "" {
+		return nil, fmt.Errorf("AccessKeyID or SecretAccessKey is empty")
+	}
+
+	queryParams := url.Values{
+		"Action":        {connectAction},
+		"User":          {userName},
+		"X-Amz-Expires": {strconv.FormatInt(int64(tokenValiditySeconds), 10)},
+	}
+
+	authURL := url.URL{
+		Host:     clusterName,
+		Scheme:   "http",
+		Path:     "/",
+		RawQuery: queryParams.Encode(),
+	}
+
+	req, err := http.NewRequest(http.MethodGet, authURL.String(), nil)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return &AuthTokenGenerator{
+		serviceName: serviceName,
+		region:      cfg.Region,
+		req:         req,
+		credentials: credentials,
+		signer:      v4.NewSigner(),
+	}, nil
```
Code Climate is complaining because there are too many return statements. Is there a better pattern that allows the individual errors to still be preserved. The function is already pretty small so it feels a little weird to break it into smaller functions.

willwill96 commented:
Is there any possibility of this being reviewed/merged?

tuunit commented:
Hi @willwill96, we are only one or two maintainers checking the PRs in our free time. Therefore it often takes quite some time before get around to these things. In the case of your change, we don't have an AWS account to validate the changes in your PR therefore we usually look at other topics first. I'll try to make some time for it :)

rfahsel3 commented:
Thank you very much @tuunit . Your message makes a lot of sense.

JoelSpeed commented on pkg/apis/options/options.go:
```diff
@@ -156,6 +156,10 @@ func NewFlagSet() *pflag.FlagSet {
 	flagSet.Bool("redis-use-cluster", false, "Connect to redis cluster. Must set --redis-cluster-connection-urls to use this feature")
 	flagSet.StringSlice("redis-cluster-connection-urls", []string{}, "List of Redis cluster connection URLs (eg redis://[USER[:PASSWORD]@]HOST[:PORT]). Used in conjunction with --redis-use-cluster")
 	flagSet.Int("redis-connection-idle-timeout", 0, "Redis connection idle timeout seconds, if Redis timeout option is non-zero, the --redis-connection-idle-timeout must be less then Redis timeout option")
+	flagSet.Bool("redis-aws-use-iam-auth", false, "Use AWS IAM authentication for Redis. Must set --redis-aws-service-name, --redis-aws-cluster-name, & --redis-aws-username to use this feature")
+	flagSet.String("redis-aws-service-name", "", "AWS service name for Redis IAM authentication -  `elasticache` or  `memorydb`")
+	flagSet.String("redis-aws-cluster-name", "", "AWS cluster name for Redis IAM authentication")
+	flagSet.String("redis-aws-username", "", "AWS username for Redis IAM authentication")
```
Do we need these as flags/could they be added only to the alpha/structured configuration?

JoelSpeed commented on pkg/sessions/redis/aws-iam/auth.go:
```diff
@@ -0,0 +1,98 @@
+	hexEncodedSHA256EmptyString = "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855"
+)
+
+type IAMTokenGenerator struct {
```
Nit missing comment on exported struct

Does it actually need to be exported? Could be behind an interface?

JoelSpeed commented on pkg/sessions/redis/aws-iam/auth.go:
```diff
@@ -0,0 +1,98 @@
+	signer              *v4.Signer
+}
+
+func New(serviceName, clusterName, userName string) (*IAMTokenGenerator, error) {
```
Missing comment on exported function

JoelSpeed commented on pkg/sessions/redis/aws-iam/auth.go:
```diff
@@ -0,0 +1,98 @@
+		return "", fmt.Errorf("AWS IAM request signing failed - %v", err)
+	}
+
+	signedURL = strings.Replace(signedURL, "http://", "", 1)
```
Can you add a comment explaining why we do this?

JoelSpeed commented on pkg/sessions/redis/redis_store.go:
```diff
@@ -170,12 +171,39 @@ func buildStandaloneClient(opts options.RedisStoreOptions) (Client, error) {
+		}
+		return opts.AWSUsername, token
+	}
+	// AWS services has a max connection lifetime of 12 hours. This is set to 11 hours to give some buffer time
```
I assume there's no action needed and the connection refreshes itself automatically after this period?

JoelSpeed reviewed (commented):
Is there any way we can add testing locally to prove this is working? Is there a service that we could use to mock the AWS auth?

willwill96 commented on pkg/sessions/redis/aws-iam/auth_test.go:
```diff
@@ -0,0 +1,35 @@
+	"github.com/stretchr/testify/require"
+)
+
+func TestAWSIAMTokenGenerator(t *testing.T) {
```
I saw another OSS project add a similar unit test when they added IAM auth for postgres:
https://github.com/authzed/spicedb/pull/1858/files

willwill96 commented:
Hi @JoelSpeed,

Thank you for taking a look at our changes. I think that I have addressed your comments to the best of my ability, but happy to make additional changes if necessary. 

Additionally, I have sent an email to you & @tuunit containing a demo of the IAM auth changes, which I hope will inspire a bit more confidence in these changes. 

If you have any additional feedback or questions on these changes, I'm happy to help where I can.

Best regards.

rfahsel3 commented:
Hi @tuunit ! I saw you made some contributions to this MR. Is this ready to merge?

rfahsel3 commented:
@tuunit and @JoelSpeed , I'm sorry to be annoying, but we are really trying to figure out if we are going to fork this or not. We have a pretty hard deadline at the end of the month. I feel like we were making good progress by sending the video and working through the comments. Do we foresee this being merged in the next week, or should we plan on forking the repo? We just need to know for our planning.
